lib/src/builder/parsing/ast_util.dart

```diff
@@ -128,11 +128,17 @@ extension MetadataHelper on AnnotatedNode {
     }
 
     final segments = fullName.split('.');
-    return segments.lastWhere((segment) {
-      final withoutSpecialPrefixes = segment.replaceFirst(RegExp(r'^[_$]+'), '');
-      if (withoutSpecialPrefixes.isEmpty) return false;
-      return withoutSpecialPrefixes[0] == withoutSpecialPrefixes[0].toUpperCase();
-    }, orElse: () => segments.last);
+    return segments.lastWhere(_looksLikeAClassName, orElse: () => segments.last);
+  }
+
+  /// Returns whether the first character of [name] that isn't a `$` or `_` is capitalized.
+  static bool _looksLikeAClassName(String name) {
+    for (var i = 0; i < name.length; i++) {
+      final character = name[i];
+      if (character == r'$' || character == '_') continue;
+      return character == character.toUpperCase();
+    }
+    return false;
```

Comment:
When I was profiling, I micro-optimized this since I had a hunch that this RegExp (which tend to not have great performance in the Dart VM) was inefficient, and IIRC it resulted in a 2% speed up to the benchmark 😄.

Thankfully I stopped trying to optimize things before I got too far down that rabbit hole 🙂